extension/partiql-extension-ion/src/boxed_ion.rs

```diff
@@ -531,6 +545,142 @@ enum BoxedIonValue {
+
+        let result = if l.is_null() && r.is_null() {
+            NULLS_EQUAL
+        } else {
+            match (lty, rty) {
+                (IonType::Float, IonType::Float) => {
+                    let (l, r) = (l.as_float().unwrap(), r.as_float().unwrap());
+                    if l.is_nan() && r.is_nan() {
+                        NAN_EQUAL
+                    } else {
+                        l == r
+                    }
+                }
+
+                (IonType::List, IonType::List) => {
+                    let (ls, rs) = (l.as_list().unwrap(), r.as_list().unwrap());
+                    l.annotations().eq(r.annotations())
+                        && NullableEq::eq(&wrap_seq(ls), &wrap_seq(rs)) == Value::Boolean(true)
+                }
+                (IonType::SExp, IonType::SExp) => {
+                    let (ls, rs) = (l.as_sexp().unwrap(), r.as_sexp().unwrap());
+                    l.annotations().eq(r.annotations())
+                        && NullableEq::eq(&wrap_seq(ls), &wrap_seq(rs)) == Value::Boolean(true)
+                }
+
+                (IonType::Struct, IonType::Struct) => {
+                    let (ls, rs) = (l.as_struct().unwrap(), r.as_struct().unwrap());
+                    l.annotations().eq(r.annotations())
+                        && NullableEq::eq(&wrap_struct(ls), &wrap_struct(rs))
+                            == Value::Boolean(true)
+                }
+
+                _ => l == r,
+            }
+        };
+
+        Value::Boolean(result)
+    }
+
+    #[inline(always)]
+    fn eqg(&self, rhs: &Self) -> Value {
+        let wrap = IonEqualityValue::<'_, true, { NAN_EQUAL }, _>;
+        NullableEq::eq(&wrap(self.0), &wrap(rhs.0))
+    }
+}
+
+impl<'a, const NULLS_EQUAL: bool, const NAN_EQUAL: bool> NullableEq
+    for IonEqualityValue<'a, NULLS_EQUAL, NAN_EQUAL, Sequence>
+{
+    fn eq(&self, other: &Self) -> Value {
+        let wrap = IonEqualityValue::<'a, { NULLS_EQUAL }, { NAN_EQUAL }, _>;
+        let (l, r) = (self.0, other.0);
+        let l = l.iter().map(wrap);
+        let r = r.iter().map(wrap);
+        let res = l.zip(r).all(|(l, r)| l.eqg(&r) == Value::Boolean(true));
+        Value::Boolean(res)
+    }
+
+    #[inline(always)]
+    fn eqg(&self, rhs: &Self) -> Value {
+        let wrap = IonEqualityValue::<'_, true, { NAN_EQUAL }, _>;
+        NullableEq::eq(&wrap(self.0), &wrap(rhs.0))
+    }
+}
+
+impl<'a, const NULLS_EQUAL: bool, const NAN_EQUAL: bool> NullableEq
+    for IonEqualityValue<'a, NULLS_EQUAL, NAN_EQUAL, Struct>
+{
+    fn eq(&self, other: &Self) -> Value {
+        let wrap = IonEqualityValue::<'a, { NULLS_EQUAL }, { NAN_EQUAL }, _>;
+        let (l, r) = (self.0, other.0);
+        let l = l.iter().map(|(s, elt)| (s, wrap(elt)));
+        let r = r.iter().map(|(s, elt)| (s, wrap(elt)));
+        let res = l
+            .zip(r)
+            .all(|((ls, lelt), (rs, relt))| ls == rs && lelt.eqg(&relt) == Value::Boolean(true));
```

Comment:
Seems like this equality is ordered? So `` `{a: 1, b: 2}` `` would not equal `` `{b: 2, a: 1}` ``?

Should the Ion equality follow PartiQL semantics and be unordered?

Reply:
Aligned with PartiQL tuple equality.